src/script/scriptBuilder.zig

```diff
@@ -0,0 +1,79 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+const Script = @import("lib.zig").Script;
+pub const engine = @import("engine.zig");
+const Opcode = @import("./opcodes/constant.zig").Opcode;
+const ScriptNum = Script.ScriptNum;
+const asBool = Script.asBool;
+const asInt = Script.asInt;
+const testing = std.testing;
+///ScriptBuilder is a library to generate easier scripts, useful for faster testing
+pub const ScriptBuilder = struct {
+    /// Number of opcodes in the script(this is not a necessary field and may be removed later)
+    opcodeCount: u8,
+
+    /// Dynamic array holding the opcodes
+    opcodes: []u8,
+
+    ///Memory Allocator
+    allocator: Allocator,
+
+    /// Initialize a new ScriptBuilder
+    ///
+    /// # Arguments
+    /// - `allocator`: Memory allocator for managing engine resources
+    ///
+    /// # Returns
+    /// - `!ScriptBuilder`: use with `try`
+    pub fn new(allocator: std.mem.Allocator) !ScriptBuilder {
+        return ScriptBuilder{
+            .opcodeCount = 0,
+            .opcodes = try allocator.alloc(u8, 0),
+            .allocator = allocator,
+        };
+    }
+
+    ///Push an OPcode to the ScriptBuilder
+    pub fn pushOpcode(self: *ScriptBuilder, op: Opcode) !*ScriptBuilder {
+        // Resize the array to add one more element
+        const newOpcodes = try self.allocator.realloc(self.opcodes, self.opcodeCount + 1);
+
+        newOpcodes[self.opcodeCount] = op.toBytes();
+        self.opcodes = newOpcodes;
+        self.opcodeCount += 1;
+        return self;
+    }
+
+    ///Push an Int to the SciptBuilder
+    /// OP_1...OP_16
+    pub fn pushInt(self: *ScriptBuilder, num: u8) !*ScriptBuilder {
+        const newOpcodes = try self.allocator.realloc(self.opcodes, self.opcodeCount + 1);
+        const op = Opcode.pushOpcode(num);
+        newOpcodes[self.opcodeCount] = op.toBytes();
+        self.opcodes = newOpcodes;
+        self.opcodeCount += 1;
+        return self;
+    }
+
```

Comment:
YOu are missing an addData fn
```zig
pub fn addData(self: *ScriptBuilder, data: []u8) !*ScriptBuilder {
// Pseudo code: 
If (data.len == 0 or data.len == 1 && data[0] == 0) {
 // push OP_0
} else if data.len == 1 && data[0] <= 16
  // Use the OP_N opcode, like this Opcode.OP_1 - 1 + num 
} else if data.len == 1 && data[0] == 0x81 {
 // push OP_1NEGATE
} 

// Now do the same king of check in order to use if possible the opcodes beetwin 0x01-0x4b , if not, the smallest OP_PUSHDATA opcode possible

```

Reply:
Regarding the following code, it seems if I pass 6 or 0x06 to the function, it takes it as numeric 6(I presume because 0x06 and 6 are treated as the same)

```
   pub fn addData(self: *ScriptBuilder, data: []u8) !*ScriptBuilder {
        if (data.len == 0 or data.len == 1 and data[0] == 0) {
            _ = try self.addOpcode(Opcode.OP_0);
        } else if (data.len == 1 and data[0] <= 16) {
            const op = Opcode.OP_1.toBytes() - 1 + data[0];
            try self.script.append(op);
        } else if (data.len == 1 and data[0] == 0x81) {
            _ = try self.addOpcode(Opcode.OP_1NEGATE);
        } else if (data.len > 1 and data[0] >= 0x01 and data[0] <= 0x4b) {
            _ = try self.script.append(data[0]);
        }
        return self;
    }
    ```

Reply:
> Do we need pushing OP_1 to OP_16 in this function?

Yeah we do need, coz we are trying to minimize the script length. So it will take one byte (OP_16) instead of two (PUSH_DATA_1, 0x6).

>  We already have a dedicated func pushInt for that
`pushInt` is not an opcode. It's an internal method of our stack